File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        p = l1
        q = l2
        carry = 0
        dummyhead = ListNode()
        cur = dummyhead
        while(p != None or q != None):
            if p != None:
                p_val = p.val
                p = p.next
            else:
                p_val = 0
            if q != None:
                q_val = q.val
                q = q.next
            else:
                q_val = 0  
            cur.next =ListNode((p_val + q_val + carry)%10)
            carry = (p_val + q_val + carry)/10 
            cur = cur.next
        if carry > 0:
            cur.next = ListNode(carry)
        return dummyhead.next
                
        
            
            
            
            
        

File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p1 = l1;
        ListNode p2 = l2;
        int carry = 0;
        ListNode dumb = new ListNode(-1);
        ListNode p3 = dumb;
        while(l1 != null || l2 != null){
            int l1_val = 0;
            int l2_val = 0;
            if (l1 != null){
                l1_val = l1.val;
                l1 = l1.next;
            }
            if (l2 != null){
                l2_val = l2.val;
                l2 = l2.next;
            }
            int add_res = l1_val + l2_val + carry;
            carry = add_res / 10;
            add_res = add_res%10;
            ListNode temp = new ListNode(add_res);
            p3.next = temp;
            p3 = p3.next;
        }
        if (carry > 0){
            ListNode temp = new ListNode(carry);
            p3.next = temp;
        }
        return dumb.next;
    }
}